File: runtime/scripts/care_router.py
"""
CareRouter — Intent-aware model selection for CareSupport.

Routes each inbound message to the right Claude model based on complexity
and safety-criticality. Zero extra API calls — pure pattern matching.

Tiers:
  fast     → Haiku   (greetings, schedule, check-ins, general)
  reason   → Sonnet  (medication changes, onboarding, multi-member coordination)
  critical → Opus    (emergencies, escalation triggers)
"""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass
class RouteResult:
    tier: str       # "fast" | "reason" | "critical"
    model: str      # Anthropic model ID
    intent: str     # classified intent for logging
    reason: str     # why this tier was chosen


MODELS = {
    "fast": "claude-haiku-4-5-20251001",
    "reason": "claude-sonnet-4-6",
    "critical": "claude-opus-4-6",
}

TIER_ORDER = ["fast", "reason", "critical"]


# ─── Pattern Definitions ─────────────────────────────────────────────────

_EMERGENCY = re.compile(
    r"\b(911|emergency|ambulance|chest pain|can'?t breathe|"
    r"choking|seizure|unconscious|unresponsive|bleeding|overdose|"
    r"heart attack|stroke|passed out|not breathing)\b|"
    r"\bfell\b(?!\s*(asleep|behind|short|silent|quiet|flat|in love))",
    re.IGNORECASE,
)

_ESCALATION = re.compile(
    r"\b(missed (his|her|the|a) (med|medication|pill|dose)|"
    r"hasn'?t taken|didn'?t take (his|her|the)|"
    r"no one (is|was) (here|available|scheduled)|"
    r"no coverage|gap in coverage|nobody showed|"
    r"hasn'?t shown up|didn'?t show up)\b",
    re.IGNORECASE,
)

_MEDICATION_CHANGE = re.compile(
    r"\b(change|stop|start|switch|adjust|increase|decrease|new|discontinue|hold|pause|resume) "
    r"(his|her|the|a|my)?\s*(\w+\s+){0,3}(med|medication|prescription|dosage|dose|pill)\b|"
    r"\b(med|medication|prescription|dosage|dose) (change|adjustment|update)\b",
    re.IGNORECASE,
)

_ONBOARDING = re.compile(
    r"\b(new (caregiver|member|person|helper|aide)|"
    r"add (someone|a member|a caregiver|a helper|them|her|him)|"
    r"join(ing)? the (team|circle|care)|"
    r"set(ting)? (up|me up)|"
    r"sign(ing)? up)\b",
    re.IGNORECASE,
)

_MULTI_MEMBER = re.compile(
    r"\b(tell|message|text|contact|reach out to|let .+ know|notify)\b.*"
    r"\b(and|also|both|everyone|all|the team)\b",
    re.IGNORECASE,
)


# ─── Router ───────────────────────────────────────────────────────────────

def route(message: str, member: dict | None = None) -> RouteResult:
    """Classify message intent and select model tier.

    Priority order — first match wins:
      1. Emergency keywords → critical
      2. Escalation triggers → critical
      3. Medication change → reason
      4. Onboarding signals → reason
      5. Multi-member coordination → reason
      6. Everything else → fast
    """
    msg = message.strip()

    if _EMERGENCY.search(msg):
        return RouteResult("critical", MODELS["critical"], "EMERGENCY", "emergency keywords detected")

    if _ESCALATION.search(msg):
        return RouteResult("critical", MODELS["critical"], "ESCALATION", "escalation trigger detected")

    if _MEDICATION_CHANGE.search(msg):
        return RouteResult("reason", MODELS["reason"], "MEDICATION_CHANGE", "medication change request")

    if _ONBOARDING.search(msg):
        return RouteResult("reason", MODELS["reason"], "ONBOARDING", "new member onboarding")

    if _MULTI_MEMBER.search(msg):
        return RouteResult("reason", MODELS["reason"], "MULTI_MEMBER", "multi-member coordination")

    return RouteResult("fast", MODELS["fast"], "GENERAL", "default routing")


def fallback_chain(start_model: str) -> list[str]:
    """Return models to try: start_model, then each higher tier."""
    chain = [start_model]
    start_tier = next((t for t, m in MODELS.items() if m == start_model), None)
    if start_tier and start_tier in TIER_ORDER:
        idx = TIER_ORDER.index(start_tier)
        for tier in TIER_ORDER[idx + 1:]:
            chain.append(MODELS[tier])
    return chain

File: convex/lib/pipeline/careRouter.ts
import type { Intent, RouteResult, RouteTier } from "./types";

export const MODELS: Record<RouteTier, string> = {
  fast: "claude-sonnet-4-6",
  reason: "claude-sonnet-4-6",
  critical: "claude-opus-4-6",
};

export const TIER_ORDER: RouteTier[] = ["fast", "reason", "critical"];

const EMERGENCY = /\b(911|emergency|ambulance|chest pain|can't breathe|not breathing|stroke|heart attack|unconscious|unresponsive|seizure|overdose)\b/i;
const MEDICATION_CHANGE = /\b(change|stop|start|switch|adjust|increase|decrease|new|discontinue|hold|pause|resume).{0,30}\b(med|medication|prescription|dosage|dose|pill)\b|\b(med|medication|prescription|dosage|dose)\s+(change|update|adjustment)\b/i;
const BILLING = /\b(price|pricing|cost|charge|billing|bill|free|paid|plan)\b/i;

const PATTERNS: Array<{
  pattern: RegExp;
  tier: RouteTier;
  intent: Intent;
  reason: string;
}> = [
  { pattern: EMERGENCY, tier: "critical", intent: "EMERGENCY", reason: "emergency keywords detected" },
  { pattern: MEDICATION_CHANGE, tier: "reason", intent: "MEDICATION_CHANGE", reason: "medication request detected" },
  { pattern: BILLING, tier: "fast", intent: "BILLING", reason: "billing question detected" },
];

export function route(message: string): RouteResult {
  const msg = message.trim();

  for (const { pattern, tier, intent, reason } of PATTERNS) {
    if (pattern.test(msg)) {
      return { tier, model: MODELS[tier], intent, reason };
    }
  }

  return { tier: "fast", model: MODELS.fast, intent: "GENERAL", reason: "default routing" };
}

export function fallbackChain(startModel: string): string[] {
  const chain = [startModel];
  const startTier = (Object.entries(MODELS) as [RouteTier, string][])
    .find(([, model]) => model === startModel)?.[0];

  if (startTier) {
    const index = TIER_ORDER.indexOf(startTier);
    for (const tier of TIER_ORDER.slice(index + 1)) {
      if (!chain.includes(MODELS[tier])) {
        chain.push(MODELS[tier]);
      }
    }
  }

  return chain;
}
